File: 0042-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& a) {
    int n = a.size();
    std::vector<int> v_leftMax(n);
    std::vector<int> v_rightMax(n);

    v_leftMax[0] = a[0];
    v_rightMax[n-1] = a[n-1];

    for(int i = 1 ; i < n ; i++){
      v_leftMax[i] = max(v_leftMax[i-1],a[i]);
    }

    for (int i = n - 2 ; i > -1 ; --i){
      v_rightMax[i] = max(v_rightMax[i+1],a[i]);
    }
    

    int water_stored = 0 ;
    for(int i = 0 ; i < n ; i++){
      water_stored += abs( min(v_leftMax[i],v_rightMax[i]) - a[i]);
    }
    return water_stored ;
    }
};

File: 0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int n = height.length ;

        int[] leftMax = new int[n];
        leftMax[0] = height[0];
        int[] rightMax = new int[n];
        rightMax[n-1] = height[n-1];

        IntStream.range(1,n).forEach( i -> leftMax[i] = Math.max(leftMax[i-1],height[i]) );
        IntStream.range(0,n-1).map(i-> n-i-2).forEach(i-> rightMax[i] = Math.max(rightMax[i+1],height[i]));
        return IntStream.range(0,n).map(i-> Math.abs(Math.min(leftMax[i],rightMax[i]) - height[i])).sum();
    }
}
